packages/zoe/zoe.js

```diff
@@ -42,137 +35,238 @@ const makeZoe = async (additionalEndowments = {}) => {
-        const assays = readOnlyState.getAssays(instanceHandle);
-        const labels = readOnlyState.getLabelsForInstanceHandle(instanceHandle);
-        const extentOpsArray = readOnlyState.getExtentOpsArrayForInstanceHandle(
+        const { assays } = instanceTable.get(instanceHandle);
+        const unitOpsArray = assayTable.getUnitOpsForAssays(assays);
+        const extentOpsArray = assayTable.getExtentOpsForAssays(assays);
+        const offerRecord = {
           instanceHandle,
-        );
-
-        const { offerHandle } = escrowEmptyOffer(
-          adminState.recordOffer,
+          payoutRules: makeEmptyPayoutRules(unitOpsArray),
+          exitRule: { kind: 'onDemand' },
           assays,
-          labels,
-          extentOpsArray,
-        );
-        return offerHandle;
-      },
-      /**
-       *  The contract can also create a real offer and get the
-       *  associated offerHandle, bypassing the escrow receipt
-       *  creation. This allows the contract to make offers on the
-       *  users' behalf, as happens in the `addLiquidity` step of the
-       *  `autoswap` contract.
-       */
-      escrowOffer: async (offerRules, offerPayments) => {
-        const { offerHandle } = await escrowOffer(
-          adminState.recordOffer,
-          adminState.recordAssay,
-          offerRules,
-          offerPayments,
-        );
+          units: unitOpsArray.map(unitOps => unitOps.empty()),
+          extents: extentOpsArray.map(extentOps => extentOps.empty()),
+        };
+        const offerHandle = offerTable.create(offerRecord);
+        payoutMap.init(offerHandle, makePromise());
         return offerHandle;
       },
 
       /**
-       * Burn the escrowReceipt ERTP payment using the escrowReceipt
-       * assay. Burning in ERTP also validates that the alleged payment was
-       * produced by the assay, and returns the
-       * units of the payment.
-       *
-       * This method also checks if the offer has been completed and
-       * errors if it has, so that a smart contract doesn't continue
-       * thinking that the offer is live. Additionally, we record that
-       * the escrowReceipt is used in this particular contract.
-       *
-       * @param  {object} escrowReceipt - an ERTP payment
-       * representing proof of escrowing specific assets with Zoe.
+       * The contract can also make an real offer with offerRules and
+       * ERTP payments. Autoswap uses this method to introduce newly
+       * minted liquidity tokens to Zoe.
        */
+      escrowOffer: (offerRules, offerPayments) => {
+        const { assays } = instanceTable.get(instanceHandle);
+        const offerImmutableRecord = {
+          instanceHandle,
+          payoutRules: offerRules.payoutRules,
+          exitRule: offerRules.exitRule,
+          assays,
+          units: undefined,
+          extents: undefined,
+        };
+        const offerHandle = offerTable.create(offerImmutableRecord);
```

Comment:
Same issue as with `assayTable`. You're entering the offer into the table before it's ready. If you postpone until ready, you'll also be able to use correct values for the `units` and `extents`. Only then does the record uphold the intended invariants. Until then we should keep this out of shared tables.

Reply:
I've changed the code so that we wait until the deposit succeeds before creating the offer record.